### PR TITLE
Add satisfaction survey performance metrics to dashboard

### PR DESCRIPTION
**Pull request by george-bit**
## Context

In order to make our positive survey result more accessible they should be added to the feature metric dashboard.

## Changes proposed in this pull request

image

## Guidance to review

Test in support App and QA to check values are representative. Any thoughts RE the naming on the dashboard front end?

## Link to Trello card

https://trello.com/c/XdeMPLhi/3520-display-csat-with-performance-metrics

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/views/support_interface/performance/feature_metrics_dashboard.html.erb [hunk 1]
```diff
@@ -408,6 +408,32 @@
         </div>
       </div>
     </div>
+    <div class="govuk-grid-column-one-half">
+      <div id="headline-stat" class="govuk-!-margin-bottom-4">
```
Comment (paulrobertlloyd) [inline app/views/support_interface/performance/feature_metrics_dashboard.html.erb:412]: Can we remove this `id`? There should only be one unique `id` per page, and I note that this is being repeated for each set of stats. I’ll make a separate PR to address this (and simplify the markup a bit, too)

```suggestion
      <div class="govuk-!-margin-bottom-4">
```

**Comment by paulrobertlloyd** [comment]
You might want to review and merge this PR https://github.com/DFE-Digital/apply-for-teacher-training/pull/5123 and rebase into this one.

File: app/models/satisfaction_survey_feature_metrics.rb [hunk 2]
```diff
@@ -12,13 +12,31 @@ def response_rate(
     success_count / (fail_count + success_count)
   end
 
+  def positive_feedback_rate(
+    start_time,
+    end_time = Time.zone.now.end_of_day
+  )
+    feedback_count = application_forms_with_feedback(start_time, end_time).count.to_f
+    positive_feedback_count = application_forms_with_positive_feedback(start_time, end_time).count.to_f
+    return nil if feedback_count.zero?
```
Comment (frankieroberto) [inline app/models/satisfaction_survey_feature_metrics.rb:21]: could move this up a line to save a query..? 😄

File: app/models/satisfaction_survey_feature_metrics.rb [hunk 3]
```diff
@@ -42,6 +60,16 @@ def application_forms_with_feedback(start_time, end_time)
       .where(recruitment_cycle_year: RecruitmentCycle.current_year)
   end
 
+  def application_forms_with_positive_feedback(start_time, end_time)
+    ApplicationForm
+      .where(
+        recruitment_cycle_year: RecruitmentCycle.current_year,
+      )
+      .where('application_forms.submitted_at BETWEEN ? AND ?', start_time, end_time)
```
Comment (frankieroberto) [inline app/models/satisfaction_survey_feature_metrics.rb:68]: I think this should work?:

```suggestion
      .where(application_forms: {submitted_at: start_time..end_time})
```

File: app/models/satisfaction_survey_feature_metrics.rb [hunk 4]
```diff
@@ -42,6 +60,16 @@ def application_forms_with_feedback(start_time, end_time)
...
+      .where(feedback_satisfaction_level: 'very_satisfied').or(ApplicationForm.where(feedback_satisfaction_level: 'satisfied'))
```
Comment (frankieroberto) [inline app/models/satisfaction_survey_feature_metrics.rb:69]: I think this should generate a `WHERE x IN (a,b)` query, which does the same thing as an OR:

```suggestion
      .where(feedback_satisfaction_level: ['very_satisfied', 'satisfied'])
```

File: app/views/support_interface/performance/feature_metrics_dashboard.html.erb [hunk 5]
```diff
@@ -408,6 +408,32 @@
...
+        <%= render SupportInterface::TileComponent.new(
+          count: @dashboard.read_metric(:satisfaction_survey_positive_feedback_rate),
+          label: 'Satisfaction Survey Positive Feedback',
```
Comment (frankieroberto) [inline app/views/support_interface/performance/feature_metrics_dashboard.html.erb:415]: Suggestion:

```suggestion
          label: 'satisfied or very satisfied',
```

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 2 above)
Comment (george-bit) [reply app/models/satisfaction_survey_feature_metrics.rb:21]: Good idea ;)

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 3 above)
Comment (george-bit) [reply app/models/satisfaction_survey_feature_metrics.rb:68]: will give it a go, was cribbing from Steve's sql

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 4 above)
Comment (george-bit) [reply app/models/satisfaction_survey_feature_metrics.rb:69]: I'll take a 👀

File: app/models/satisfaction_survey_feature_metrics.rb [hunk 6]
```diff
@@ -42,6 +60,16 @@ def application_forms_with_feedback(start_time, end_time)
...
+      .where(recruitment_cycle_year: RecruitmentCycle.current_year)
```
Comment (stevehook) [inline app/models/satisfaction_survey_feature_metrics.rb:70]: I don't think this query is quite right, it gives me the following SQL (with some sample dates):
```
SELECT application_forms.*
FROM application_forms
WHERE (application_forms.recruitment_cycle_year = 2021 AND
  (application_forms.submitted_at BETWEEN '2021-06-30 12:11:42.072318' AND '2021-07-08 12:11:42.072536')
  AND application_forms.feedback_satisfaction_level = 'very_satisfied'
  OR application_forms.feedback_satisfaction_level = 'satisfied')
AND application_forms.recruitment_cycle_year = 2021;
```

which is logically...

```
SELECT application_forms.*
FROM application_forms
WHERE (application_forms.recruitment_cycle_year = 2021 AND
  ((application_forms.submitted_at BETWEEN '2021-06-30 12:11:42.072318' AND '2021-07-08 12:11:42.072536')
  AND application_forms.feedback_satisfaction_level = 'very_satisfied')
  OR application_forms.feedback_satisfaction_level = 'satisfied')
AND application_forms.recruitment_cycle_year = 2021;
```

so you can end up with satisfaction rates > 100%

Comment (stevehook) [reply app/models/satisfaction_survey_feature_metrics.rb:70]: I think you need something like this...

```
    ApplicationForm
      .where(feedback_satisfaction_level: 'very_satisfied').or(ApplicationForm.where(feedback_satisfaction_level: 'satisfied'))
      .where(recruitment_cycle_year: RecruitmentCycle.current_year)
      .where('application_forms.submitted_at BETWEEN ? AND ?', start_time, end_time)
      .where(recruitment_cycle_year: RecruitmentCycle.current_year)
```

(the bracketing around `OR` that you end up with isn't very obvious - at least not to me).

**Review by stevehook — COMMENTED** [review]
Looks good. There's one gotcha that I almost missed.

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 4 above)
Comment (stevehook) [reply app/models/satisfaction_survey_feature_metrics.rb:69]: :+1: it will, and it will get around the issue that I highlighted with brackets and `OR`.

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 6 above)
Comment (stevehook) [reply app/models/satisfaction_survey_feature_metrics.rb:70]: or better still use an `IN` as @frankieroberto suggests

Comment (george-bit) [reply app/models/satisfaction_survey_feature_metrics.rb:70]: okay cool so something like this @stevehook @frankieroberto 

`  def application_forms_with_positive_feedback(start_time, end_time)
    ApplicationForm
      .where(feedback_satisfaction_level: %w[very_satisfied satisfied])
      .where(recruitment_cycle_year: RecruitmentCycle.current_year)
      .where(application_forms: { submitted_at: start_time..end_time })
      .where(recruitment_cycle_year: RecruitmentCycle.current_year)
  end`

Comment (frankieroberto) [reply app/models/satisfaction_survey_feature_metrics.rb:70]: @george-bit  Looks good to me, although there's a duplicate line. Depending on preferred code style, could even combine the `.where` statements:

```ruby
def application_forms_with_positive_feedback(start_time, end_time)
    ApplicationForm
      .where(
        feedback_satisfaction_level: %w[very_satisfied satisfied],
        recruitment_cycle_year: RecruitmentCycle.current_year,
        submitted_at: start_time..end_time
      )
  end
  ```

**Comment by george-bit** [comment]
All comments addressed, Paul am I okay to merge  #5123 ?

File: app/models/satisfaction_survey_feature_metrics.rb (hunk 6 above)
Comment (george-bit) [reply app/models/satisfaction_survey_feature_metrics.rb]: This is much neater thanks.

**Comment by paulrobertlloyd** [comment]
> All comments addressed, Paul am I okay to merge #5123 ?

I have already merged it in, so a rebase should address any outstanding issues.

**Comment by george-bit** [comment]
Cool, have just rebased 👍

**Review by paulrobertlloyd — APPROVED** [review]
